### DISCUSSION TITLE
C#网络数据流处理

### DISCUSSION THREAD

**Discussion by lizayu666** [BYJRK/DotNet-Discussions#168] [C# 技术交流]
在处理网络数据流时，往往数据流的编码不是固定的，很可能是未知的。如果要对数据流中某一段英语字母进行替换，比如把数据流中最后的 (0.02 secs). 这个去除掉，用C#则需要从网络数据流的bytes数组通过一个编码转换成字符串，然后用正则或者字符串函数去处理。但就像刚刚说到，bytes数组转为字符串在未知编码的情况下就会出现错误。
python3是基于bytes类型编程，python在处理数据流时就不需要把bytes数据根据某个编码直接转为字符串后处理。python的字符串函数可以直接对bytes数据进行操作，而无需把bytes根据某个编码转为字符串，比如：

```python
import re

# 假设networkdata是从网络中获取的数据包，数据包中的编码是未知的
# 需求只需要替换(0.02 secs).结尾的英文数字字符
networkdata = b"The data obtained from the Internet (0.02 secs).\r\n"  #   不要尝试用decode之类对某个编码转为str类型
if networkdata.strip().endswith(b" (0.02 secs)."):
    networkdata = re.sub(rb"\s\([0-9+.]+\ssecs\)\.\r\n$", b'\r\n', networkdata, 1)
print(networkdata)  # 可以把networkdata数据发送出去
```

上面python就可以很方便的在不知道编码的情况下，直接对bytes类型进行操作。请问下，C#有没有类似在编码未知的情况下，对数据流进行上述python类似的操作？

**Comment by seishinkouki** [comment]
python中 =b'abc'的声明方式实际上默认了是ascii的字符串, 你如果传一个utf-16 encode的bytes后面的endswith(b" (0.02 secs).")啥的就不管用了, 
![image](https://github.com/BYJRK/DotNet-Discussions/assets/8842321/d006b4d6-92e3-49cd-8f92-cd530358bd9a)

**Reply by lizayu666** [reply]
我的上述代码
networkdata = b"The data obtained from the Internet (0.02 secs).\r\n
只是打一个比方，假设从网络上获取到这些信息。实际上真正项目，从网络上获取的bytes数据，我们是不知道具体编码的

**Comment by seishinkouki** [comment]
编码未知的情况下, 是可以根据一定的字节流猜测编码的

**Comment by BYJRK** [comment]
这个可以用上吗？根据一部分字节判断编码：

https://github.com/yinyue200/ude

**Reply by lizayu666** [reply]
我故意用gbk（gb2312）编码的字符写到一个文件中，用此github提供的代码运行，显示探测失败

```
string filename = "filename.txt";  // gbk编码的中文字
using (var fs = File.OpenRead(filename))
{
    Ude.CharsetDetector cdet = new Ude.CharsetDetector();
    cdet.Feed(fs);
    cdet.DataEnd();
    if (cdet.Charset != null)
    {
        Console.WriteLine("Charset: {0}, confidence: {1}",
             cdet.Charset, cdet.Confidence);
    }
    else
    {
        Console.WriteLine("Detection failed.");
    }
}
```
